translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -9441,15 +9478,155 @@ Value *SPIRVToLLVM::transGLSLExtInst(SPIRVExtInst *ExtInst,
     // Inverse of square matrix
     return getBuilder()->CreateMatrixInverse(Args[0]);
 
+  case GLSLstd450Modf: {
+    // Split input into fractional and whole number parts.
+    Value *WholeNum =
+        getBuilder()->CreateUnaryIntrinsic(Intrinsic::trunc, Args[0]);
+    Value *Fract = getBuilder()->CreateFSub(Args[0], WholeNum);
+    getBuilder()->CreateStore(WholeNum, Args[1]);
+    return Fract;
+  }
+
+  case GLSLstd450ModfStruct: {
+    // Split input into fractional and whole number parts.
+    Value *WholeNum =
+        getBuilder()->CreateUnaryIntrinsic(Intrinsic::trunc, Args[0]);
+    Value *Fract = getBuilder()->CreateFSub(Args[0], WholeNum);
+    Value *Result = UndefValue::get(transType(ExtInst->getType()));
+    Result = getBuilder()->CreateInsertValue(Result, Fract, 0);
+    Result = getBuilder()->CreateInsertValue(Result, WholeNum, 1);
+    return Result;
+  }
+
+  case GLSLstd450FMin:
+  case GLSLstd450NMin: {
+    // FP minimum
+    FastMathFlags FMF;
+    FMF.setNoNaNs();
+    CallInst *Result = getBuilder()->CreateBinaryIntrinsic(Intrinsic::minnum,
+                                                           Args[0], Args[1]);
+    Result->setFastMathFlags(FMF);
+    return Result;
+  }
+
+  case GLSLstd450UMin: {
+    // Unsigned integer minimum
+    Value *Cmp = getBuilder()->CreateICmpULT(Args[1], Args[0]);
+    return getBuilder()->CreateSelect(Cmp, Args[1], Args[0]);
+  }
+
+  case GLSLstd450SMin: {
+    // Signed integer minimum
+    Value *Cmp = getBuilder()->CreateICmpSLT(Args[1], Args[0]);
+    return getBuilder()->CreateSelect(Cmp, Args[1], Args[0]);
+  }
+
+  case GLSLstd450FMax:
+  case GLSLstd450NMax: {
+    // FP maximum
+    FastMathFlags FMF;
+    FMF.setNoNaNs();
+    CallInst *Result = getBuilder()->CreateBinaryIntrinsic(Intrinsic::maxnum,
+                                                           Args[0], Args[1]);
+    Result->setFastMathFlags(FMF);
+    return Result;
+  }
+
+  case GLSLstd450UMax: {
+    // Unsigned integer maximum
+    Value *Cmp = getBuilder()->CreateICmpULT(Args[1], Args[0]);
+    return getBuilder()->CreateSelect(Cmp, Args[0], Args[1]);
+  }
+
+  case GLSLstd450SMax: {
+    // Signed integer maximum
+    Value *Cmp = getBuilder()->CreateICmpSLT(Args[1], Args[0]);
+    return getBuilder()->CreateSelect(Cmp, Args[0], Args[1]);
+  }
+
+  case GLSLstd450FClamp:
+  case GLSLstd450NClamp:
+    // FP clamp
+    return getBuilder()->CreateFMed3(Args[0], Args[1], Args[2]);
```

Comment:
please set nan flag for GLSLstd450FClamp

Reply:
Actually I created a new CreateFClamp to use here, and set FMF appropriately for it depending on whether it is FClamp or NClamp. I discovered that using CreateFMed3 is bad for a datatype that does not support the fmed3 instruction, because it does 2x min and 2x max, whereas for a clamp you only need 1x min and 1x max.